src/k2/dto/Expression.cpp

```diff
@@ -0,0 +1,298 @@
+                if (rec.schema->fields[i].name == val.fieldName) {
+                    sfieldIndex = i;
+                    nullLast = rec.schema->fields[i].nullLast;
+                    type = rec.schema->fields[i].type;
+                    break;
+                }
+            }
+        }
+    }
+    Value& val;
+    SKVRecord& rec;
+    FieldType type = FieldType::NOT_KNOWN;
+    bool nullLast = false;
+    int sfieldIndex = -1;
+
+    template <typename T>
+    std::optional<std::tuple<bool, T>> get() {
+        if (TToFieldType<T>() != type) {
+            throw TypeMismatchException();
+        }
+        if (val.fieldName.empty()) {
+            T result{};
+            if (val.literal.read(val)) {
+                return std::make_tuple(nullLast, std::move(result));
+            }
+            return std::nullopt;
+        }
+        auto&& opt = rec.deserializeField<T>(sfieldIndex);
+        if (!opt) {
+            return std::nullopt;
+        }
+        return std::make_tuple(nullLast, std::move(opt).value());
+    }
+};
+
+template <typename T1, typename T2, typename = void>
+struct is_comparable : std::false_type {};
+
+template <typename T1, typename T2>
+struct is_comparable<T1, T2,
+                     decltype(
+                         (std::declval<T1>() == std::declval<T2>()) &&
+                             (std::declval<T1>() < std::declval<T2>()) &&
+                             (std::declval<T1>() > std::declval<T2>()),
+                         void())>
+    : std::true_type {};
+
+template <typename T1, typename T2>
+std::enable_if_t<is_comparable<T1, T2>::value, int>
+compare(T1& a, T2& b) {
+    if (std::forward<T1>(a) == std::forward<T2>(b)) return 0;
+    if (std::forward<T1>(a) > std::forward<T2>(b)) return 1;
+    return -1;
+}
+
+template <typename T1, typename T2>
+std::enable_if_t<!is_comparable<T1, T2>::value, int>
+compare(T1&, T2&) {
+    throw TypeMismatchException();
+}
+
+
+template <typename T1, typename T2>
+std::enable_if_t<!is_comparable<T1, T2>::value, int>
+compareOptionals(std::optional<std::tuple<bool, T1>>&, std::optional<std::tuple<bool,T2>>&) {
+    throw TypeMismatchException();
+}
+
+template <typename T1, typename T2>
+std::enable_if_t<is_comparable<T1, T2>::value, int>
+compareOptionals(std::optional<std::tuple<bool, T1>>& a, std::optional<std::tuple<bool,T2>>& b) {
+    if (!a && !b) {
+        // NULLs of compatible types compare as equal
+        return 0;
+    }
+
+    // one of the operands may be null. Compare based on nullLast flag, regardless of type
+    if (!a) {
+        return std::get<0>(*b) ? 1 : -1;
+    }
+    else if (!b) {
+        return std::get<0>(*a) ? -1 : 1;
+    }
+
+    // operands are non-null and of same type. Compare their values
+    if (std::get<1>(*a) == std::get<1>(*b)) {
+        return -1;
+    }
+    if (std::get<1>(*a) > std::get<1>(*b)) {
+        return 1;
+    }
+    return -1;
+}
+
+#define CAST_APPLY_VALUE(func, a, ...)              \
```

Comment:
This is fairly generic now which is great. As one of my next tasks I can find all the places in the codebase that has a switch like this and consolidate.

Reply:
Yes, in particular, i made it work off of .type field so that we could also consolidate with the SKV's DO_ON_NEXT_RECORD_FIELD macro